ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -575,6 +579,113 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
     );
   }
 
+  private void verifyJoin(final Join root, final PlanNode left, final PlanNode right) {
+    if (left.getNodeOutputType() == DataSourceType.KSTREAM) {
+      if (right.getNodeOutputType() == DataSourceType.KTABLE) {
+
+        // stream-table join detected
+
+        if (root.getInfo().getType().equals(JoinType.OUTER)) {
+          throw new KsqlException("Invalid join type:"
+              + " full-outer join not supported for stream-table joins.");
+        }
+
+        if (joinOnNonKeyAttribute(root.getInfo().getRightJoinExpression(), right)) {
+          throw new KsqlException("Invalid join condition:"
+              + " stream-table joins require to join on the table's primary key.");
+        }
+      }
+
+      // stream-stream join detected: nothing to verify
+
+    } else {
+      if (right.getNodeOutputType() == DataSourceType.KSTREAM) {
+        throw new KsqlException("Invalid join order:"
+            + " table-stream joins are not supported; only stream-table joins.");
+      }
+
+      // table-table join detected
+
+      if (joinOnNonKeyAttribute(root.getInfo().getRightJoinExpression(), right)) {
+        throw new KsqlException("Invalid join condition:"
+            + " table-table joins require to join on the primary key of the right input table."
+        );
+      }
+
+      if (joinOnNonKeyAttribute(root.getInfo().getLeftJoinExpression(), left)) {
+        // after we lift the n-way join restriction, we should be able to support FK-joins
+        // at any level in the join tree, even after we add right-deep/bushy join tree support,
+        // because a FK-join output table has the same PK as its left input table
+        throw new KsqlException("Invalid join condition:"
+            + " foreign-key table-table joins are not supported.");
```

Comment:
nit: (to clarify that there is nothing fundamentally unsupported about this, just that we haven't added support yet)
```suggestion
            + " foreign-key table-table joins are not yet supported.");
```

Reply:
Well, we add them shortly anyway...